### PR TITLE
feat: enable transparent cargo commands

### PR DESCRIPTION
**Pull request by jdx**
## Keep typing `cargo`

This PR makes mbx work transparently with normal Cargo commands. Run setup once, then use the commands you already know:

```sh
cargo build
cargo test
cargo clippy
```

Cargo still owns its command line. Aliases, installed Cargo subcommands, `+toolchain`, help, version output, and arguments after `--` continue to work normally.

## Install with mise

```sh
mise use --global --postinstall "mbx setup --yes" mr-boxington
```

> [!NOTE]
> This installation method requires mise 2026.8.15 or newer.

The hook exports the source configuration as `MISE_CONFIG_FILE`, so `mbx setup --yes` activates the Cargo shim in the configuration that contains the tool:

- With `mise use --global …`, activation is global.
- With project-local `mise use …`, activation is limited to that project.

This depends on [jdx/mise#12603](https://github.com/jdx/mise/pull/12603) for scoped `env._.path` updates and `MISE_CONFIG_FILE`, plus the separate `mise use --postinstall` change.

## Install without mise

```sh
cargo install mbx --locked
mbx setup
```

When mise is active in the current shell, `mbx setup` recommends the config that defines `mr-boxington`, then the nearest project config, and finally the global config. It prompts before using that recommendation; `mbx setup --yes` accepts it. A postinstall hook still uses the exact `MISE_CONFIG_FILE` supplied by mise.

Without an active mise shell, setup prints the shell-specific command for prepending the shim directory to `PATH`. It supports sh-compatible shells, fish, Nushell, and PowerShell, and does not edit shell startup files.

## What setup does\n\nSetup also configures rust-analyzer’s `check.overrideCommand` in the matching global or project scope. It invokes the stable Cargo shim by absolute path, putting editor checks in the same cache and machine-wide scheduler pool even when the editor did not inherit mise’s `PATH`. Existing check settings are preserved, and uninstall removes only the command owned by mbx.\n\n`mbx setup` installs an executable named `cargo` in mbx's platform data directory and activates that directory ahead of the real Cargo on `PATH`.

On Unix, the shim is a symlink to the installed `mbx`. On Windows, `cargo.exe` forwards to the active `mbx` on `PATH`, with the setup-time path as a fallback. Upgrading `mbx` therefore does not leave Cargo running stale mbx code. There is no separate update command.

```sh
mbx setup --status
mbx setup --uninstall
mbx doctor
```

`mbx doctor` reports whether the shim is installed, current, and first on `PATH`. A missing shim is only a warning because explicit `mbx build`, `mbx test`, and similar commands remain a supported zero-config alternative.

Uninstall removes activation from the selected mise scope while leaving the shared shim bytes in place in case another scope still references them.

## Runtime behavior

When invoked as `cargo`, mbx resolves the real Cargo while excluding its own directory. Workspace compilation commands enter a normal mbx cache session. Clearly non-compiling operations pass directly to Cargo, and unknown commands remain Cargo's so aliases and plugins continue to work.

To bypass mbx for one invocation in a POSIX shell:

```sh
MBX_DISABLE=1 cargo build
```

In PowerShell, set `$env:MBX_DISABLE = "1"`, run Cargo, then remove it with `Remove-Item Env:MBX_DISABLE`.

Users who do not want setup can continue to run `mbx build`, `mbx test`, or `mbx +1.91 check` directly. mbx's own commands remain under the `mbx` executable, including `tui`, `cache`, `gc`, `doctor`, and diagnostics.

## Validation

- Rust unit and integration tests
- setup lifecycle and transparent-Cargo Bats tests, including cross-checkout rustdoc restoration through plain `cargo doc`
- an isolated-home mise E2E covering the documented local and global `--postinstall` commands, Cargo resolution, `mbx doctor`, cross-checkout cache reuse, and uninstall
- rust-analyzer setup, status, preservation, and uninstall coverage\n- Windows PowerShell shim coverage
- lint, generated CLI documentation, API compatibility, fuzz, supply-chain, and performance checks

The real mise E2E intentionally runs the documented commands without feature detection or skipping. CI will remain red at that test while released mise rejects `--postinstall` or the scoped collection-update flags.

*AI-assisted — Tool: Codex; model: GPT-5; version: unavailable.*

---

> [!NOTE]
> **Medium Risk**
> Setup mutates mise config, PATH resolution, and editor rust-analyzer settings; a misconfigured or stale shim could route builds incorrectly, though passthrough and `MBX_DISABLE` limit blast radius.
> 
> **Overview**
> **Replaces the hidden `rustc-wrapper` setup with a public `mbx setup` flow** so normal `cargo build` / `test` / `clippy` can use the cache after one-time configuration, while `mbx …` remains the zero-config alternative.
> 
> `mbx setup` now installs a stable **`cargo` shim** under the platform data directory (shell launcher on Unix, tracking shim on Windows), optionally prepends that directory via **mise `env._.path`** (`--global` / `--local` / `--yes`, including postinstall with `MISE_CONFIG_FILE`), or prints manual **PATH** instructions without touching shell startup files. It also sets **rust-analyzer** `check.overrideCommand` to the shim (preserving existing check settings) and drops **`--update`** in favor of refresh-on-install.
> 
> A new **cargo shim runtime** (`shim.rs`, early dispatch in `main`) resolves real Cargo by excluding the shim directory, passes through non-build commands and **`MBX_DISABLE`**, and runs compile commands through the existing mbx session; explicit `mbx` cargo paths call **`prepare_explicit_cargo`** to avoid shim recursion. **`mbx doctor`** now reports shim install/current/active-on-PATH instead of Cargo config wrapper state.
> 
> Docs, README, CLI reference, CI (**mise-action** on non-Windows), and Bats/PowerShell tests cover the lifecycle, mise scopes, cross-checkout caching, and Windows shim behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 362ef428d9e0fa9b354d69d977f7aff23cf895c2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

- **New Features**
  - Added `mbx setup` to enable standard `cargo` commands to run through mbx.
  - Supports global or project-local activation, status checks, confirmation prompts, and uninstalling activation.
  - Cargo shim support now works across Unix and Windows environments.
  - Configures rust-analyzer to use the active Cargo shim.
  - `mbx`-prefixed Cargo commands remain available as a zero-configuration alternative.
- **Documentation**
  - Updated installation, setup, migration, usage, and troubleshooting guidance.
  - Added dedicated CLI documentation for `mbx setup`.
- **Bug Fixes**
  - Improved detection of missing, outdated, inactive, or non-executable Cargo shims.
  - Prevented recursive shim invocation during diagnostics and tooling commands.

### CODE REVIEW

File: crates/mbx/src/cli.rs [hunk 1]
```diff
@@ -441,144 +480,235 @@ fn setup_paths() -> Result<(PathBuf, PathBuf)> {
...
+    Select::new("Where should plain cargo commands use mbx?")
+        .option(
+            DemandOption::new(MiseScope::Global)
+                .label("Everywhere mise is active")
+                .selected(true),
+        )
+        .option(DemandOption::new(MiseScope::Local).label("Only in this project"))
+        .option(DemandOption::new(MiseScope::None).label("Create the shim without activating it"))
+        .run()
+        .map_err(Into::into)
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli.rs]: Fixed in e559af0. Interactive setup detects whether mise exists before offering activation scopes; without mise it installs the shim and prints manual PATH instructions.

File: crates/mbx/src/cli.rs [hunk 2]
```diff
@@ -591,6 +721,142 @@ fn same_file_contents(left: &Path, right: &Path) -> Result<bool> {
...
+}
+
+fn cargo_proxy_passthrough(arguments: &[OsString]) -> bool {
+    if arguments.iter().any(|argument| {
+        matches!(
+            argument.to_str(),
+            Some("--version" | "-V" | "--help" | "-h")
+        )
+    }) {
+        return true;
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli.rs]: Fixed in e559af0. Help/version detection stops at the Cargo -- separator, with a regression assertion for cargo run -- --help.

File: crates/mbx/src/cli.rs [hunk 3]
```diff
@@ -416,10 +433,32 @@ fn validate_prefetch_config(config: &Config) -> Result<()> {
...
+    }
+}
+
+fn setup(args: &SetupArgs, action: SetupAction) -> Result<ExitCode> {
     let executable = std::env::current_exe().wrap_err("failed to locate the mbx executable")?;
-    let (install_dir, config_path) = setup_paths()?;
-    setup_at_action(&executable, &install_dir, &config_path, action)
+    let (install_dir, cargo_config) = setup_paths()?;
+    let scope = setup_scope(args)?;
+    setup_at_action(&executable, &install_dir, &cargo_config, &scope, action)
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli.rs]: Fixed in e559af0. Unscoped status, update, and uninstall no longer invoke the installation scope prompt; explicit --global/--local and postinstall file scopes remain scope-aware.

File: crates/mbx/src/cli.rs [hunk 4]
```diff
@@ -591,6 +721,142 @@ fn same_file_contents(left: &Path, right: &Path) -> Result<bool> {
...
+    }
+    let name = if cfg!(windows) { "cargo.exe" } else { "cargo" };
+    for directory in std::env::split_paths(path) {
+        if Some(directory.as_path()) == current_dir {
+            continue;
+        }
+        let candidate = directory.join(name);
+        if candidate.is_file() {
+            return Ok(candidate.into_os_string());
+        }
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli.rs]: Fixed in e559af0. Cargo resolution compares shim directories case-insensitively on Windows, and argv0 shim identification is case-insensitive there as well. Windows-only regression assertions cover both.

File: crates/mbx/src/cli.rs [hunk 5]
```diff
@@ -591,6 +721,142 @@ fn same_file_contents(left: &Path, right: &Path) -> Result<bool> {
...
+                | "logout"
+                | "owner"
+                | "package"
+                | "publish"
+                | "yank"
+                | "locate-project"
+                | "metadata"
+                | "generate-lockfile"
+        )
+    )
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli.rs]: Fixed in e559af0. cargo install and uninstall, along with other clearly non-compiling built-ins, pass directly to real Cargo. Cached Cargo aliases remain covered by the behavioral suite.

File: crates/mbx/src/cli.rs [hunk 6]
```diff
@@ -441,144 +480,235 @@ fn setup_paths() -> Result<(PathBuf, PathBuf)> {
...
+            .as_deref()
+            .and_then(|shell| Path::new(shell).file_stem())
+            .and_then(OsStr::to_str);
+        println!("prepend the Cargo shim to PATH in your shell:");
+        match shell {
+            Some("fish") => println!("  fish_add_path {path}"),
+            Some("nu") | Some("nushell") => {
+                println!("  $env.PATH = ($env.PATH | prepend '{path}')")
+            }
+            _ => println!("  export PATH=\"{path}:$PATH\""),
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli.rs]: Already fixed in e559af0: print_manual_activation now takes install_dir directly and prints its absolute path for every shell.

File: crates/mbx/src/cli.rs [hunk 7]
```diff
@@ -591,6 +721,142 @@ fn same_file_contents(left: &Path, right: &Path) -> Result<bool> {
...
+
+/// Run an invocation received through the persistent Cargo shim.
+pub fn run_cargo_shim() -> Result<ExitCode> {
+    let current = std::env::current_exe().wrap_err("failed to locate the Cargo shim")?;
+    let real_cargo = resolve_real_cargo(&current)?;
+    let arguments = std::env::args_os().skip(1).collect::<Vec<_>>();
+    if mbx_disabled() || cargo_proxy_passthrough(&arguments) {
+        return run_real_cargo(&real_cargo, &arguments);
+    }
+    let arguments = strings(&arguments)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli.rs]: Fixed in 9df5561. Invalid UTF-8 arguments now fall back to real Cargo with the original OsString bytes intact.

File: e2e-win/CargoDispatch.Tests.ps1 [hunk 8]
```diff
@@ -67,4 +67,32 @@ mbx-probe = "new --vcs none"
...
+        $originalLocalAppData = $env:LOCALAPPDATA
+        $originalPath = $env:PATH
+        $originalDisable = $env:MBX_DISABLE
+        try {
+            $env:LOCALAPPDATA = Join-Path $script:TestRoot 'local-app-data'
+            $output = & $mbx setup 2>&1 | Out-String
+            $LASTEXITCODE | Should -Be 0 -Because $output
+
+            $shimDir = Join-Path $env:LOCALAPPDATA 'mbx\bin'
+            Test-Path -LiteralPath (Join-Path $shimDir 'cargo.exe') | Should -BeTrue
```
[the remark this answers is not in the document]
Comment (jdx) [reply e2e-win/CargoDispatch.Tests.ps1]: Fixed in 9df5561. The PowerShell E2E assertion now derives the shim directory from LocalApplicationData via Environment.GetFolderPath, matching dirs::data_local_dir.

File: README.md [hunk 9]
```diff
@@ -83,6 +83,9 @@ archives cover both libc variants on x86-64 and ARM64, plus Apple Silicon and
 Windows x86-64 and ARM64.
 Every release includes `SHA256SUMS`.
 
+After installing from an archive, run `mbx setup`. Without mise it prints the
+appropriate `PATH` command and does not edit your shell startup files.
```
[the remark this answers is not in the document]
Comment (jdx) [reply README.md]: Fixed in 9df5561. Archive instructions invoke the extracted binary by its explicit Unix or Windows path before PATH activation.

File: test/setup.bats [hunk 10]
```diff
@@ -5,15 +5,13 @@ setup() {
...
   _common_setup
-
-  if [[ -n "$rustup_home" ]]; then
-    export RUSTUP_HOME="$rustup_home"
-  fi
+  if [[ -n "$rustup_home" ]]; then export RUSTUP_HOME="$rustup_home"; fi
   export CARGO_HOME="$BATS_TEST_TMPDIR/cargo-home"
+  export XDG_DATA_HOME="$BATS_TEST_TMPDIR/data-home"
   export MBX_CACHE_DIR="$BATS_TEST_TMPDIR/mbx-cache"
+  export MBX_SHIM_DIR="$XDG_DATA_HOME/mbx/bin"
```
[the remark this answers is not in the document]
Comment (jdx) [reply test/setup.bats:14]: Fixed in 9df5561. The shared Bats setup now unsets MISE_CONFIG_FILE; scope-specific tests set it explicitly when needed.

File: crates/mbx/src/cli.rs [hunk 11]
```diff
@@ -441,144 +480,235 @@ fn setup_paths() -> Result<(PathBuf, PathBuf)> {
...
+            .and_then(|shell| Path::new(shell).file_stem())
+            .and_then(OsStr::to_str);
+        println!("prepend the Cargo shim to PATH in your shell:");
+        match shell {
+            Some("fish") => println!("  fish_add_path {path}"),
+            Some("nu") | Some("nushell") => {
+                println!("  $env.PATH = ($env.PATH | prepend '{path}')")
+            }
+            _ => println!("  export PATH=\"{path}:$PATH\""),
+        }
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli.rs]: Fixed in e559af0. Manual shell instructions now receive the absolute install directory; the portable tilde form is used only for mise env._.path.

File: crates/mbx/src/doctor.rs [hunk 12]
```diff
@@ -238,74 +238,105 @@ fn reflink_check(cache_dir: &Path) -> Check {
...
+            .as_deref()
+            .map(|cargo| cargo.display().to_string())
+            .unwrap_or_else(|| "nothing".into());
+        return Check::warn(
             "setup",
-            format!("configured wrapper is missing: {}", expected_shim.display()),
+            format!(
+                "Cargo shim is current but not active; PATH resolves cargo to {resolved}; run `mbx setup` for activation instructions"
+            ),
+        );
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/doctor.rs:291]: Fixed in b0fa2a7. Doctor now gives the exact shim directory to prepend to PATH instead of recommending a refresh-only setup command.

File: crates/mbx/src/cli/setup.rs [hunk 13]
```diff
@@ -215,3 +481,79 @@ pub(super) fn same_file_contents(left: &Path, right: &Path) -> Result<bool> {
...
+pub(crate) fn cargo_shim_is_current(executable: &Path, shim: &Path) -> Result<bool> {
+    #[cfg(windows)]
+    if let Some(install_dir) = shim.parent()
+        && let Some(configured) = configured_cargo_shim_target(install_dir)
+    {
+        return Ok(same_path(&configured, executable)
+            || cargo_shim_target(install_dir)
+                .is_some_and(|target| same_path(&target, executable)));
+    }
+    same_file_contents(executable, shim)
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/setup.rs]: Fixed in 51ca64e. The recorded target is now parsed independently of whether it still exists: runtime fallback only executes it when it is a file, while status and doctor can still recognize the sidecar and compare the current active mbx from PATH. The Windows E2E now deletes the recorded-version target logically, verifies setup status against the active PATH mbx, and invokes cargo.exe through that path.

File: crates/mbx/src/cli/setup.rs [hunk 14]
```diff
@@ -19,191 +36,601 @@ pub(super) struct SetupArgs {
...
+        SetupAction::Install => {
+            let mut command = toml_edit::Array::new();
+            command.extend(expected);
+            document["check"]["overrideCommand"] = toml_edit::value(command);
+            crate::util::write_atomic(path, document.to_string().as_bytes())?;
+            println!(
+                "rust-analyzer background checks now run through mbx: {}",
+                path.display()
+            );
+            Ok(ExitCode::SUCCESS)
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/setup.rs:375]: Fixed in 5a88b8d. Setup now treats any existing rust-analyzer check settings as user-owned and leaves the file byte-for-byte unchanged, including command and features.

File: crates/mbx/src/cli/setup.rs [hunk 15]
```diff
@@ -19,191 +36,601 @@ pub(super) struct SetupArgs {
...
+                global()
+            } else {
+                let directory = path.parent().ok_or_else(|| {
+                    eyre::eyre!("mise configuration path has no parent: {}", path.display())
+                })?;
+                Ok(crate::util::workspace_root(directory).join(RUST_ANALYZER_CONFIG_FILE))
             }
-            eyre::bail!("mbx setup is not installed; run `mbx setup` first");
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/setup.rs:282]: Fixed in 5a88b8d. Project-scoped setup now prefers the active Cargo workspace, and unscoped status/uninstall reuse an existing workspace rust-analyzer.toml before falling back to global configuration. Regression coverage uses a repo-level mise.toml above a nested crate.

[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/setup.rs:282]: Fixed in 5a88b8d. Project-scoped setup now prefers the active Cargo workspace, and unscoped status/uninstall reuse an existing workspace rust-analyzer.toml before falling back to global configuration. Regression coverage uses a repo-level mise.toml above a nested crate.

File: crates/mbx/src/cli/setup.rs [hunk 16]
```diff
@@ -19,191 +36,619 @@ pub(super) struct SetupArgs {
...
+        (root.join("Cargo.toml").is_file() || root.join("Cargo.lock").is_file())
+            .then(|| root.join(RUST_ANALYZER_CONFIG_FILE))
+    };
+    match scope {
+        MiseScope::Global => global(),
+        MiseScope::None if action != SetupAction::Install => {
+            let local = local()?;
+            if local.is_file() { Ok(local) } else { global() }
+        }
+        MiseScope::None => global(),
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/setup.rs]: Fixed in ec63e0c. `MiseScope::None` now always selects the global rust-analyzer config for install, status, and uninstall. Project config selection remains explicit through `--local` or the mise-provided file scope, so all lifecycle actions are symmetric.

File: crates/mbx/src/cli/shim.rs [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+                | "publish"
+                | "install"
+                | "uninstall"
+                | "yank"
+                | "locate-project"
+                | "metadata"
+                | "generate-lockfile"
+        )
+    )
+}
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/shim.rs:169]: Fixed in a45ac86. Bare Cargo invocations and recognized non-compiling commands (including clean, config, fmt, and vendor) now bypass mbx before workspace probing/session setup; unknown names remain session-eligible because they may be compiling Cargo aliases. Added focused classification coverage, and the full mbx test suite plus clippy pass.

File: crates/mbx/src/cli/setup.rs [hunk 18]
```diff
@@ -19,191 +36,610 @@ pub(super) struct SetupArgs {
...
+            return Ok(ExitCode::SUCCESS);
+        }
+        SetupAction::Install => {
+            std::fs::create_dir_all(install_dir)?;
+            crate::session::install_shim_named(
+                executable,
+                install_dir,
+                CARGO_SHIM_STEM,
+                crate::session::ShimLink::Tracking,
+            )?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/setup.rs:188]: Fixed in 88e0b8e. Unix now installs a stable executable launcher instead of a symlink to a versioned mbx path. It uses the setup-time executable while present, then resolves the active mbx from PATH after an upgrade removes that version. Bats coverage deletes the setup-time binary and verifies plain cargo still dispatches successfully; setup status and doctor cover launcher integrity and executability.

File: crates/mbx/src/cli/shim.rs [hunk 19]
```diff
@@ -0,0 +1,453 @@
...
+        if skip_value {
+            skip_value = false;
+            return None;
+        }
+        let argument = argument.to_str()?;
+        if matches!(argument, "--color" | "--config" | "-Z") {
+            skip_value = true;
+            return None;
+        }
+        (!argument.starts_with('-') && !argument.starts_with('+')).then_some(argument)
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/shim.rs:130]: Fixed in 439540e. Cargo -C is now treated as a value-taking global option before command classification, with a regression assertion that cargo -C project clean bypasses the mbx session.
